homeassistant/components/climate/__init__.py

```diff
@@ -558,17 +564,75 @@ async def async_turn_on(self) -> None:
             if mode not in self.hvac_modes:
                 continue
             await self.async_set_hvac_mode(mode)
-            break
+            return
+
+        raise NotImplementedError
+
+    @property
+    def _is_turn_on_supported(self) -> bool:
+        if (
+            type(self).async_turn_on is not ClimateEntity.async_turn_on
+            or type(self).turn_on is not ClimateEntity.turn_on
+        ):
+            return True
+
+        if len(self.hvac_modes) == 2 and HVACMode.OFF in self.hvac_modes:
+            return True
+
+        for mode in (HVACMode.HEAT_COOL, HVACMode.HEAT, HVACMode.COOL):
+            if mode in self.hvac_modes:
+                return True
+
+        return False
+
+    def turn_off(self) -> None:
+        """Turn the entity off."""
+        raise NotImplementedError
 
     async def async_turn_off(self) -> None:
         """Turn the entity off."""
-        if hasattr(self, "turn_off"):
+        if type(self).turn_off is not ClimateEntity.turn_off:
             await self.hass.async_add_executor_job(self.turn_off)
             return
 
         # Fake turn off
         if HVACMode.OFF in self.hvac_modes:
             await self.async_set_hvac_mode(HVACMode.OFF)
+            return
+
+        raise NotImplementedError
+
+    @property
+    def _is_turn_off_supported(self) -> bool:
+        if (
+            type(self).async_turn_off is not ClimateEntity.async_turn_off
+            or type(self).turn_off is not ClimateEntity.turn_off
+        ):
+            return True
+
+        if HVACMode.OFF in self.hvac_modes:
+            return True
+
+        return False
+
+    def toggle(self) -> None:
+        """Toggle the entity."""
+        raise NotImplementedError
+
+    async def async_toggle(self) -> None:
+        """Toggle the entity."""
+        if type(self).toggle is not ClimateEntity.toggle:
+            await self.hass.async_add_executor_job(self.toggle)
+            return
+
+        if not (self._is_turn_on_supported and self._is_turn_off_supported):
+            raise NotImplementedError
+
+        # We assume that since turn_off is supported, HVACMode.OFF is as well.
```

Comment:
This seems like a valid assumption, we should document `turn_off`, `turn_on` and `toggle` here https://developers.home-assistant.io/docs/core/entity/climate and for turn_off mention the HVAC mode should be set to `HVACMode.Off`

Reply:
Do you agree with https://github.com/home-assistant/core/pull/100418#discussion_r1353410885? If so, I could add something like that to the docs.